File: js/src/parser.ts
import matter from "gray-matter";
import type { Metadata } from "./types.js";
import { ParseError } from "./errors.js";

/**
 * Parse a SKILL.md file into its metadata and body.
 */
export function parseSkillMd(content: string): {
  metadata: Metadata;
  body: string;
} {
  const { data, content: body } = matter(content);

  if (!data.name || typeof data.name !== "string") {
    throw new ParseError('SKILL.md frontmatter must include a "name" field');
  }
  if (!data.description || typeof data.description !== "string") {
    throw new ParseError('SKILL.md frontmatter must include a "description" field');
  }

  const metadata: Metadata = {
    name: data.name,
    description: data.description,
  };

  if (data.license != null) {
    metadata.license = String(data.license);
  }

  if (data.compatibility != null) {
    metadata.compatibility = String(data.compatibility);
  }

  const tools = data["allowed-tools"] ?? data.allowedTools;
  if (tools != null) {
    if (!Array.isArray(tools)) {
      throw new ParseError('"allowed-tools" must be an array');
    }
    metadata.allowedTools = tools.map(String);
  }

  if (data.metadata != null) {
    if (typeof data.metadata !== "object" || Array.isArray(data.metadata)) {
      throw new ParseError('"metadata" must be an object');
    }
    metadata.metadata = data.metadata as Record<string, unknown>;
  }

  return { metadata, body: body.trim() };
}

File: python/src/skills_sdk/parser.py
from __future__ import annotations

import frontmatter

from .errors import ParseError
from .types import Metadata


def parse_skill_md(content: str) -> tuple[Metadata, str]:
    post = frontmatter.loads(content)
    data = post.metadata

    if not isinstance(data.get("name"), str) or not data["name"]:
        raise ParseError('SKILL.md frontmatter must include a "name" field')
    if not isinstance(data.get("description"), str) or not data["description"]:
        raise ParseError('SKILL.md frontmatter must include a "description" field')

    metadata = Metadata(name=data["name"], description=data["description"])

    if data.get("license") is not None:
        metadata.license = str(data["license"])

    if data.get("compatibility") is not None:
        metadata.compatibility = str(data["compatibility"])

    tools = data.get("allowed-tools")
    if tools is None:
        tools = data.get("allowedTools")
    if tools is not None:
        if not isinstance(tools, list):
            raise ParseError('"allowed-tools" must be an array')
        metadata.allowed_tools = [str(t) for t in tools]

    if data.get("metadata") is not None:
        if not isinstance(data["metadata"], dict):
            raise ParseError('"metadata" must be an object')
        metadata.metadata = data["metadata"]

    return metadata, post.content.strip()
